File: src/graphics/circle_shape.rs
use {
    crate::{
        ffi::graphics as ffi,
        graphics::{
            Color, Drawable, FloatRect, IntRect, RenderStates, RenderTarget, Shape, Texture,
            Transform, Transformable,
        },
        system::Vector2f,
    },
    std::{
        marker::PhantomData,
        ptr::{self, NonNull},
    },
};

/// Specialized shape representing a circle.
#[derive(Debug)]
pub struct CircleShape<'s> {
    handle: NonNull<ffi::sfCircleShape>,
    texture: PhantomData<&'s Texture>,
}

impl<'s> CircleShape<'s> {
    /// Create a new circle shape initialized with a texture
    ///
    /// # Arguments
    /// * texture - The texture to initialize the `CircleShape` with.
    #[must_use]
    pub fn with_texture(texture: &'s Texture) -> CircleShape<'s> {
        let mut shape = CircleShape::default();
        shape.set_texture(texture, true);
        shape
    }

    /// Create a new `CircleShape` and initialize it.
    ///
    /// # Arguments:
    /// * radius - The radius of the `CircleShape`
    /// * `point_count` - The number of points to define the `CircleShape`
    ///
    /// Default value on SFML are radius = 0 / pointCount = 30
    #[must_use]
    pub fn new(radius: f32, point_count: usize) -> CircleShape<'s> {
        let mut shape = CircleShape::default();
        shape.set_radius(radius);
        shape.set_point_count(point_count);
        shape
    }

    /// Set the radius of a circle
    ///
    /// # Arguments
    /// * radius - New radius of the circle
    pub fn set_radius(&mut self, radius: f32) {
        unsafe { ffi::sfCircleShape_setRadius(self.handle.as_ptr(), radius) }
    }

    /// Set the radius of a circle
    ///
    /// Return the radius of the circle
    #[must_use]
    pub fn radius(&self) -> f32 {
        unsafe { ffi::sfCircleShape_getRadius(self.handle.as_ptr()) }
    }

    /// Set the number of points of a circle
    ///
    /// # Arguments
    /// * count - New number of points of the circle
    pub fn set_point_count(&mut self, count: usize) {
        unsafe { ffi::sfCircleShape_setPointCount(self.handle.as_ptr(), count) }
    }
    pub(super) fn raw(&self) -> *const ffi::sfCircleShape {
        self.handle.as_ptr()
    }
}

impl Default for CircleShape<'_> {
    fn default() -> Self {
        let circle = unsafe { ffi::sfCircleShape_new() };
        CircleShape {
            handle: NonNull::new(circle).expect("Failed to create CircleShape"),
            texture: PhantomData,
        }
    }
}

impl Drawable for CircleShape<'_> {
    fn draw<'a: 'shader, 'texture, 'shader, 'shader_texture>(
        &'a self,
        target: &mut dyn RenderTarget,
        states: &RenderStates<'texture, 'shader, 'shader_texture>,
    ) {
        target.draw_circle_shape(self, states)
    }
}

impl Transformable for CircleShape<'_> {
    fn set_position<P: Into<Vector2f>>(&mut self, position: P) {
        unsafe { ffi::sfCircleShape_setPosition(self.handle.as_ptr(), position.into()) }
    }
    fn set_rotation(&mut self, angle: f32) {
        unsafe { ffi::sfCircleShape_setRotation(self.handle.as_ptr(), angle) }
    }
    fn set_scale<S: Into<Vector2f>>(&mut self, scale: S) {
        unsafe { ffi::sfCircleShape_setScale(self.handle.as_ptr(), scale.into()) }
    }
    fn set_origin<O: Into<Vector2f>>(&mut self, origin: O) {
        unsafe { ffi::sfCircleShape_setOrigin(self.handle.as_ptr(), origin.into()) }
    }
    fn position(&self) -> Vector2f {
        unsafe { ffi::sfCircleShape_getPosition(self.handle.as_ptr()) }
    }
    fn rotation(&self) -> f32 {
        unsafe { ffi::sfCircleShape_getRotation(self.handle.as_ptr()) }
    }
    fn get_scale(&self) -> Vector2f {
        unsafe { ffi::sfCircleShape_getScale(self.handle.as_ptr()) }
    }
    fn origin(&self) -> Vector2f {
        unsafe { ffi::sfCircleShape_getOrigin(self.handle.as_ptr()) }
    }
    fn move_<O: Into<Vector2f>>(&mut self, offset: O) {
        unsafe { ffi::sfCircleShape_move(self.handle.as_ptr(), offset.into()) }
    }
    fn rotate(&mut self, angle: f32) {
        unsafe { ffi::sfCircleShape_rotate(self.handle.as_ptr(), angle) }
    }
    fn scale<F: Into<Vector2f>>(&mut self, factors: F) {
        unsafe { ffi::sfCircleShape_scale(self.handle.as_ptr(), factors.into()) }
    }
    fn transform(&self) -> &Transform {
        unsafe { &*ffi::sfCircleShape_getTransform(self.handle.as_ptr()) }
    }
    fn inverse_transform(&self) -> &Transform {
        unsafe { &*ffi::sfCircleShape_getInverseTransform(self.handle.as_ptr()) }
    }
}

impl<'s> Shape<'s> for CircleShape<'s> {
    fn set_texture(&mut self, texture: &'s Texture, reset_rect: bool) {
        unsafe { ffi::sfCircleShape_setTexture(self.handle.as_ptr(), texture, reset_rect) }
    }
    fn disable_texture(&mut self) {
        unsafe { ffi::sfCircleShape_setTexture(self.handle.as_ptr(), ptr::null_mut(), true) }
    }
    fn set_texture_rect(&mut self, rect: IntRect) {
        unsafe { ffi::sfCircleShape_setTextureRect(self.handle.as_ptr(), rect) }
    }
    fn set_fill_color(&mut self, color: Color) {
        unsafe { ffi::sfCircleShape_setFillColor(self.handle.as_ptr(), color) }
    }
    fn set_outline_color(&mut self, color: Color) {
        unsafe { ffi::sfCircleShape_setOutlineColor(self.handle.as_ptr(), color) }
    }
    fn set_outline_thickness(&mut self, thickness: f32) {
        unsafe { ffi::sfCircleShape_setOutlineThickness(self.handle.as_ptr(), thickness) }
    }
    fn texture(&self) -> Option<&'s Texture> {
        unsafe {
            let raw = ffi::sfCircleShape_getTexture(self.handle.as_ptr());
            raw.as_ref()
        }
    }
    fn texture_rect(&self) -> IntRect {
        unsafe { ffi::sfCircleShape_getTextureRect(self.handle.as_ptr()) }
    }
    fn fill_color(&self) -> Color {
        unsafe { ffi::sfCircleShape_getFillColor(self.handle.as_ptr()) }
    }
    fn outline_color(&self) -> Color {
        unsafe { ffi::sfCircleShape_getOutlineColor(self.handle.as_ptr()) }
    }
    fn outline_thickness(&self) -> f32 {
        unsafe { ffi::sfCircleShape_getOutlineThickness(self.handle.as_ptr()) }
    }
    fn point_count(&self) -> usize {
        unsafe { ffi::sfCircleShape_getPointCount(self.handle.as_ptr()) }
    }
    fn point(&self, index: usize) -> Vector2f {
        unsafe { ffi::sfCircleShape_getPoint(self.handle.as_ptr(), index) }
    }
    fn local_bounds(&self) -> FloatRect {
        unsafe { ffi::sfCircleShape_getLocalBounds(self.handle.as_ptr()) }
    }
    fn global_bounds(&self) -> FloatRect {
        unsafe { ffi::sfCircleShape_getGlobalBounds(self.handle.as_ptr()) }
    }
}

impl<'s> Clone for CircleShape<'s> {
    /// Return a new `CircleShape` or panic if there is not enough memory
    fn clone(&self) -> CircleShape<'s> {
        let circle = unsafe { ffi::sfCircleShape_cpy(self.handle.as_ptr()) };
        CircleShape {
            handle: NonNull::new(circle).expect("Not enough memory to clone CircleShape"),
            texture: self.texture,
        }
    }
}

impl Drop for CircleShape<'_> {
    fn drop(&mut self) {
        unsafe { ffi::sfCircleShape_del(self.handle.as_ptr()) }
    }
}

File: CSFML/src/SFML/Graphics/CircleShape.cpp

//
// SFML - Simple and Fast Multimedia Library
// Copyright (C) 2007-2018 Laurent Gomila (laurent@sfml-dev.org)
//
// This software is provided 'as-is', without any express or implied warranty.
// In no event will the authors be held liable for any damages arising from the use of this software.
//
// Permission is granted to anyone to use this software for any purpose,
// including commercial applications, and to alter it and redistribute it freely,
// subject to the following restrictions:
//
// 1. The origin of this software must not be misrepresented;
//    you must not claim that you wrote the original software.
//    If you use this software in a product, an acknowledgment
//    in the product documentation would be appreciated but is not required.
//
// 2. Altered source versions must be plainly marked as such,
//    and must not be misrepresented as being the original software.
//
// 3. This notice may not be removed or altered from any source distribution.
//



// Headers

#include <SFML/Graphics/CircleShape.h>
#include <SFML/Graphics/CircleShapeStruct.h>
#include <SFML/Graphics/Color.hpp>
#include <SFML/Graphics/ConvertTransform.hpp>
#include <cstddef>



sfCircleShape* sfCircleShape_create(void)
{
    sfCircleShape* shape = new sfCircleShape;
    shape->Texture = NULL;

    return shape;
}



sfCircleShape* sfCircleShape_copy(const sfCircleShape* shape)
{


    return new sfCircleShape(*shape);
}



void sfCircleShape_destroy(sfCircleShape* shape)
{
    delete shape;
}



void sfCircleShape_setPosition(sfCircleShape* shape, sfVector2f position)
{
    shape->This.setPosition(position.x, position.y);
}



void sfCircleShape_setRotation(sfCircleShape* shape, float angle)
{
    shape->This.setRotation(angle);
}



void sfCircleShape_setScale(sfCircleShape* shape, sfVector2f scale)
{
    shape->This.setScale(scale.x, scale.y);
}



void sfCircleShape_setOrigin(sfCircleShape* shape, sfVector2f origin)
{
    shape->This.setOrigin(origin.x, origin.y);
}



sfVector2f sfCircleShape_getPosition(const sfCircleShape* shape)
{
    sfVector2f position = {0, 0};


    sf::Vector2f sfmlPos = shape->This.getPosition();
    position.x = sfmlPos.x;
    position.y = sfmlPos.y;

    return position;
}



float sfCircleShape_getRotation(const sfCircleShape* shape)
{
    return shape->This.getRotation();
}



sfVector2f sfCircleShape_getScale(const sfCircleShape* shape)
{
    sfVector2f scale = {0, 0};


    sf::Vector2f sfmlScale = shape->This.getScale();
    scale.x = sfmlScale.x;
    scale.y = sfmlScale.y;

    return scale;
}



sfVector2f sfCircleShape_getOrigin(const sfCircleShape* shape)
{
    sfVector2f origin = {0, 0};


    sf::Vector2f sfmlOrigin = shape->This.getOrigin();
    origin.x = sfmlOrigin.x;
    origin.y = sfmlOrigin.y;

    return origin;
}



void sfCircleShape_move(sfCircleShape* shape, sfVector2f offset)
{
    shape->This.move(offset.x, offset.y);
}



void sfCircleShape_rotate(sfCircleShape* shape, float angle)
{
    shape->This.rotate(angle);
}



void sfCircleShape_scale(sfCircleShape* shape, sfVector2f factors)
{
    shape->This.scale(factors.x, factors.y);
}



sfTransform sfCircleShape_getTransform(const sfCircleShape* shape)
{


    shape->Transform = convertTransform(shape->This.getTransform());
    return shape->Transform;
}



sfTransform sfCircleShape_getInverseTransform(const sfCircleShape* shape)
{


    shape->InverseTransform = convertTransform(shape->This.getInverseTransform());
    return shape->InverseTransform;
}



void sfCircleShape_setTexture(sfCircleShape* shape, const sfTexture* texture, sfBool resetRect)
{
    shape->This.setTexture(texture ? texture->This : NULL, resetRect == sfTrue);
    shape->Texture = texture;
}



void sfCircleShape_setTextureRect(sfCircleShape* shape, sfIntRect rect)
{
    shape->This.setTextureRect(sf::IntRect(rect.left, rect.top, rect.width, rect.height));
}



void sfCircleShape_setFillColor(sfCircleShape* shape, sfColor color)
{
    shape->This.setFillColor(sf::Color(color.r, color.g, color.b, color.a));
}



void sfCircleShape_setOutlineColor(sfCircleShape* shape, sfColor color)
{
    shape->This.setOutlineColor(sf::Color(color.r, color.g, color.b, color.a));
}



void sfCircleShape_setOutlineThickness(sfCircleShape* shape, float thickness)
{
    shape->This.setOutlineThickness(thickness);
}



const sfTexture* sfCircleShape_getTexture(const sfCircleShape* shape)
{


    return shape->Texture;
}



sfIntRect sfCircleShape_getTextureRect(const sfCircleShape* shape)
{
    sfIntRect rect = {0, 0, 0, 0};


    sf::IntRect sfmlRect = shape->This.getTextureRect();
    rect.left = sfmlRect.left;
    rect.top = sfmlRect.top;
    rect.width = sfmlRect.width;
    rect.height = sfmlRect.height;

    return rect;
}



sfColor sfCircleShape_getFillColor(const sfCircleShape* shape)
{
    sfColor color = {0, 0, 0, 0};


    sf::Color sfmlColor = shape->This.getFillColor();
    color.r = sfmlColor.r;
    color.g = sfmlColor.g;
    color.b = sfmlColor.b;
    color.a = sfmlColor.a;

    return color;
}



sfColor sfCircleShape_getOutlineColor(const sfCircleShape* shape)
{
    sfColor color = {0, 0, 0, 0};


    sf::Color sfmlColor = shape->This.getOutlineColor();
    color.r = sfmlColor.r;
    color.g = sfmlColor.g;
    color.b = sfmlColor.b;
    color.a = sfmlColor.a;

    return color;
}



float sfCircleShape_getOutlineThickness(const sfCircleShape* shape)
{
    return shape->This.getOutlineThickness();
}



size_t sfCircleShape_getPointCount(const sfCircleShape* shape)
{
    return shape->This.getPointCount();
}



sfVector2f sfCircleShape_getPoint(const sfCircleShape* shape, size_t index)
{
    sfVector2f point = {0, 0};


    sf::Vector2f sfmlPoint = shape->This.getPoint(index);
    point.x = sfmlPoint.x;
    point.y = sfmlPoint.y;

    return point;
}



void sfCircleShape_setRadius(sfCircleShape* shape, float radius)
{
    shape->This.setRadius(radius);
}



float sfCircleShape_getRadius(const sfCircleShape* shape)
{
    return shape->This.getRadius();
}



void sfCircleShape_setPointCount(sfCircleShape* shape, size_t count)
{
    shape->This.setPointCount(count);
}



sfFloatRect sfCircleShape_getLocalBounds(const sfCircleShape* shape)
{
    sfFloatRect rect = {0, 0, 0, 0};


    sf::FloatRect sfmlRect = shape->This.getLocalBounds();
    rect.left = sfmlRect.left;
    rect.top = sfmlRect.top;
    rect.width = sfmlRect.width;
    rect.height = sfmlRect.height;

    return rect;
}



sfFloatRect sfCircleShape_getGlobalBounds(const sfCircleShape* shape)
{
    sfFloatRect rect = {0, 0, 0, 0};


    sf::FloatRect sfmlRect = shape->This.getGlobalBounds();
    rect.left = sfmlRect.left;
    rect.top = sfmlRect.top;
    rect.width = sfmlRect.width;
    rect.height = sfmlRect.height;

    return rect;
}
